src/shipit_uplift/shipit_uplift/coverage_by_changeset_impl.py

```diff
@@ -0,0 +1,196 @@
+
+        data = r.json()
+
+        if r.status_code != requests.codes.ok:
+            if data['error']['reason'] == 'File not found in report':
+                return None
+
+            raise Exception('Can\'t load codecov.io report')
+
+        return dict([(int(l), v) for l, v in data['commit']['report']['files'][filename]['l'].items()])
+    elif COVERAGE_SERVICE == CoverageService.ACTIVEDATA:
+        print(json.dumps({
+            'from': 'coverage-summary',
+            'where': {
+                'and': [
+                    {
+                        'eq': {
+                            'source.file.name': filename,
+                        },
+                    },
+                    {
+                        'eq': {
+                            'repo.changeset.id12': changeset[:12],
+                        },
+                    },
+                ],
+            },
+        }))
+
+        r = requests.post('https://activedata.allizom.org/query', data=json.dumps({
+            'from': 'coverage-summary',
+            'where': {
+                'and': [
+                    {
+                        'eq': {
+                            'source.file.name': filename,
+                        },
+                    },
+                    {
+                        'eq': {
+                            'repo.changeset.id12': changeset[:12],
+                        },
+                    },
+                ],
+            },
+        }))
+
+        f = r.json()['source']['file']
+        uncovered = f['uncovered']
+        covered = [e['line'] for e in f['covered']]
+
+        all_data = [0] * max(uncovered, covered)
+        for c in covered:
+            all_data[c] = 1
+
+        return all_data
+    else:
+        assert False
+
+def has_coverage(changeset):
+    if COVERAGE_SERVICE == CoverageService.COVERALLS:
+        r = requests.get('https://coveralls.io/builds/%s.json' % get_github_commit(changeset))
+        return r.status_code == requests.codes.ok
+    elif COVERAGE_SERVICE == CoverageService.CODECOV:
+        r = requests.get('https://codecov.io/api/gh/marco-c/gecko-dev/commit/%s' % get_github_commit(changeset))
+        return r.status_code == requests.codes.ok
+    elif COVERAGE_SERVICE == CoverageService.ACTIVEDATA:
+        r = requests.post('https://activedata.allizom.org/query', data=json.dumps({
+            'from': 'coverage-summary',
+            'where': {
+                'and': [
+                    {
+                        'eq': {
+                            'repo.changeset.id12': changeset[:12],
+                        }
+                    }
+                ]
+            },
+            'select': {
+                'aggregate': 'count'
+            }
+        }))
+        return r.json()['data']['count'] > 0
+    else:
+        assert False
+
+
+def get_coverage_builds(changeset):
+    r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/' + changeset)
```

Comment:
I would have probably turned 'https://hg.mozilla.org/mozilla-central' into a constant at the top of the file.